ui/src/app/utils.ts

```diff
@@ -0,0 +1,18 @@
+import {Observable} from 'rxjs/Observable';
+
+// Don't retry API calls when the status code is 500 or 400.
+export function retryApi(observable: Observable<any>,
+    toRun: number): Observable<any> {
+  let numberRuns = 0;
+  return observable.retryWhen((errors) => {
+    return errors.do((e) => {
+      numberRuns++;
+      if (numberRuns === 3) {
+        throw e;
+      }
+      if (e.status === 500 || e.status === 400) {
```

Comment:
There are other status codes (e.g. 409, 403, 401, etc.) that shouldn't be retried, either.

Basically the only status code that means "retry please" is 503. I think that should be our criteria.

Reply:
I'm tempted to retry on regular 500s, too, even though technically one isn't supposed to. We have had for example SQL deadlock errors where the client code could save the user some grief by retrying on a 500.

I think a whitelist for retry codes is right though.

Reply:
I would argue the right answer there is for us to either avoid SQL deadlock errors, or return a 503 when they happen. 

500 means "we have a bug / aren't handling something right", so I'd prefer not to retry on it... we don't know the implications of retrying, since we didn't intend to return a 500 in the first place.

(When we have a bug, the user suffers, and we have to investigate and fix before they stop suffering.)

On a related note, we have a problem with retries generally (for 503s or otherwise) in that we don't generally handle the case where a client re-sends a request we already processed, and particularly when creating data / sending an email / doing something we don't want to repeat. 

So for example, @blrubenstein ran into an issue where his workspace creation logic threw an error after the Firecloud workspace was created; retry created a conflict. (Which is confusing.) That's going to be a potential problem for 503s too, which we need to handle; in the meantime it's nicer not to retry things for 500s.

Reply:
(I recognize that unhandled exceptions throw 500s, and some percentage of the time retrying will fix the issue... I just think it's better to err on the side of not retrying when you're unsure what's going on.)

Reply:
I like this. I'm not sure it's feasible return a 503 on deadlocks, but I like the idea of doing this.

Reply:
On retry handling, I filed this issue:

https://precisionmedicineinitiative.atlassian.net/browse/RW-194

Something we can consider addressing much later.